core/templates/services/prevent-page-unload-event.service.ts

```diff
@@ -25,35 +25,46 @@ import { WindowRef } from 'services/contextual/window-ref.service';
 })
 export class PreventPageUnloadEventService {
   private listenerActive: boolean;
+  validationCallback: Function;
+  _preventPageUnloadEventHandlerBind = undefined;
   constructor(private windowRef: WindowRef) {
     this.listenerActive = false;
+    this.validationCallback = undefined;
   }
 
-  addListener(): void {
+  addListener(callback: Function = () => {
+    return true;
+  }): void {
     if (this.listenerActive) {
       return;
     }
+    this.validationCallback = callback;
+    this._preventPageUnloadEventHandlerBind =
+      this._preventPageUnloadEventHandler.bind(null, this.validationCallback);
     this.windowRef.nativeWindow.addEventListener(
-      'beforeunload', this._preventPageUnloadEventHandler);
+      'beforeunload', this._preventPageUnloadEventHandlerBind, true);
     this.listenerActive = true;
   }
 
   removeListener(): void {
     this.windowRef.nativeWindow.removeEventListener(
-      'beforeunload', this._preventPageUnloadEventHandler);
+      'beforeunload', this._preventPageUnloadEventHandlerBind, true);
     this.listenerActive = false;
   }
 
   private _preventPageUnloadEventHandler(
-      e: BeforeUnloadEvent): void {
-    // The preventDefault call is used to trigger a confirmation before leaving.
-    e.preventDefault();
-    // According to the specification, to show the confirmation dialog an
-    // event handler should call preventDefault() on the event. However note
-    // that not all browsers support this method. So returnValue is also used.
-    // The exact value in returnValue is not relevant, but it needs to be set
-    // in order to trigger a confirmation before leaving.
-    e.returnValue = '';
+      validationCallback: Function, e: BeforeUnloadEvent): void {
+    if (validationCallback()) {
```

Comment:
If `validationCallback` is undefined, this will throw an error. Per a comment above it seems like this is an optional argument?

Reply:
Please fix the type definition here for `validationCallback`.

Reply:
DONE

Reply:
`validationCallback` can not be an optional argument since there is another variable `e` after it. The argument `e` is passed from the library itself, so I cannot move it behind `validationCallback`.

Reply:
Thanks for adding a comment in L39-40 that clarifies the `validationCallback` cannot be optional.

There is still a comment regarding the typedefinition of the validationCallback argument. Has that been addressed?

Reply:
> There is still a comment regarding the typedefinition of the validationCallback argument. Has that been addressed?

I don't think the type definition has to changed since it will no longer be undefined, since I'm giving it a value here: https://github.com/oppia/oppia/blob/a77d93e6e842cd0155aa50d62a57953881eca44e/core/templates/services/prevent-page-unload-event.service.ts#L41

Reply:
Considering that we know `validationCallback` will be a function that returns a boolean value would it make sense to use `() => boolean` (like it is done in `addListener`) rather than a generic `Function` type?